{% load static %}
<html>
  <head>
    <link href="http://maxcdn.bootstrapcdn.com/bootstrap/3.3.0/css/bootstrap.min.css" rel="stylesheet" id="bootstrap-css">
    <script src="http://maxcdn.bootstrapcdn.com/bootstrap/3.3.0/js/bootstrap.min.js"></script>
    <script src="https://code.jquery.com/jquery-3.5.1.js"></script>
    <script src ="https://cdn.datatables.net/1.10.23/js/jquery.dataTables.min.js"></script>
    <script src="https://cdn.datatables.net/1.10.16/js/dataTables.bootstrap4.min.js"></script>
    <script src="https://cdn.datatables.net/buttons/1.6.5/js/dataTables.buttons.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jszip/3.1.3/jszip.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/pdfmake/0.1.53/pdfmake.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/pdfmake/0.1.53/vfs_fonts.js"></script>
    <script src="https://cdn.datatables.net/buttons/1.6.5/js/buttons.html5.min.js"></script>
  
  </head>
  <body style="font-family:Arial;margin: 50px;"></body>
    <div id="source-html">
      <!-- SPECIMEN DETAILS -->
      <h4 style="background-color:#0059b3;color:White">SPECIMEN </h4>
      <br/>
      <b><label for="tumortype">Tumor Type:</b> {{ tumor_type }}</label><br/>
      <br/>
      <b><label for="tumorpurity">Tumor Purity (%):</b> </label><br/>
      <br/>
      <b><p id="tumorsample">Surgical Tumor Specimen: {{ tumor_sample }}</p></b>
      <br>
      <br>

      <h4 style="background-color:#0059b3;color:White">RESULTS </h4>
      <!-- DNA RESULTS -->
      <br/>
      <h4> DNA Sequencing results</h4>
      <p><b>•	Coding (non-synonymous) variants: </b>None.</p><br/>

        <!-- DNA Copy number changes -->
        <br/>
        <h4> DNA Copy Number Changes</h4>
        <p><b>•	Copy Number Variants: </b>None.</p><br/>

        <!-- RNA RESULTS -->
      <h4> Targeted RNA Sequencing results</h4>
      <p><b>•	The following fusions were detected in the tumor: </b>None.</p><br/>
      <br/>
      <br/>
      <br/>

      <div><p><b> Note: </b> Variants are categorized into three tiers: <br/></p></div><br>

            <div><p>Tier 1/2 - variants with strong/potential clinical significance.</p></div>
            <div><p>Tier 3 - variants with likely benign or unknown clinical significance.</p></div><br/>
     
    <h4 style="background-color:#0059b3;color:White">BACKGROUND </h4>

    <p>The Oncomine™ Focus Assay* is a multi-biomarker NGS assay that enables the detection of variants in 50 key solid tumor genes. Results are evaluated in the tumor tissue only. <br/>
       Purpose: To identify somatic mutations, focal copy number variants and fusions in the tumor that may improve diagnosis, modify therapy or allow the patient to enter ongoing trials. 
    </p>
        
    <h4 style="background-color:#0059b3;color:White">METHODS </h4>
        
    DNA and (or) RNA were extracted from Paraffin embedded tissues, were tested by next generation sequencing using the Oncomine Focus gene panel for specific mutations in 50 genes:
        
    <h5><b><u>33 Hotspot genes:</u></b></h5>
        <table border="2">
            <tr><td>AKT1</td><td>ALK</td><td>AR</td><td>BRAF</td><td>CDK4</td><td>CTNNB1</td><td>DDR2</td><td>EGFR</td><td>ERBB2</td><td>ERBB3</td></tr>
            <tr><td>ERBB4</td><td>ESR1</td><td>FGFR2</td><td>FGFR3</td><td>GNA11</td><td>GNAQ</td><td>HRAS</td><td>IDH1</td><td>IDH2</td><td>JAK1</td></tr>
            <tr><td>JAK2</td><td>JAK3</td><td>KIT</td><td>KRAS</td><td>MET</td><td>MTOR</td><td>NRAS</td><td>PDGFRA</td><td>PIK3CA</td><td>RAF1</td></tr>
            <tr><td>RET</td><td>ROS1</td><td>SMO</td></tr>   
        </table><br/>
        
    <h5><b><u>19 Copy number variants:</u></b></h5>
        <table border="2">
            <tr><td>ALK</td><td>AR</td><td>BRAF</td><td>CCDN1</td><td>CDK4</td><td>CDK6</td><td>EGFR</td><td>ERBB2</td><td>FGFR1</td><td>FGFR2</td></tr>
            <tr><td>FGFR3</td><td>FGFR4</td><td>KIT</td><td>KRAS</td><td>MET</td><td>MYC</td><td>MYCN</td><td>PDGFRA</td><td>PIK3CA</td></tr>
        </table><br/>
        
    <h5><b><u>23 Fusion drivers:</u></b></h5>
        <table border="2">
            <tr><td>ABL1</td><td>AKT3</td><td>ALK</td><td>AXL</td><td>BRAF</td><td>ERG</td><td>ETV1</td><td>ETV4</td><td>ETV5</td><td>EGFR</td></tr>
            <tr><td>ERBB2</td><td>FGFR1</td><td>FGFR2</td><td>FGFR3</td><td>MET</td><td>NTRK1</td><td>NTRK2</td><td>NTRK3</td><td>PDGFRA</td><td>PPARG</td></tr>
            <tr><td>RAF1</td><td>RET</td><td>ROS1</td>
        </table><br/>
        
    <p>A detailed list of all regions covered by the test is available upon request. The specific mutations are detected by amplification of the corresponding exons by polymerase chain reaction (PCR). The PCR product is sequenced on an Ion Torrent S5 instrument. Analysis is performed using Ion Reporter Software 5.18.</p>
        
    <h5><b>References:</b></h5>
        <p> 1. Mehrotra M, Duose DY, Singh RR, Barkoh BA, Manekia J, Harmon MA, et al. (2017)  Versatile ion S5XL sequencer for targeted next generation sequencing of solid tumors in a clinical laboratory. PLoS ONE 12(8): e0181968. Https://doi.org/10.1371/journal.pone.0181968</p>
        <p> 2. Daniel H. Hovelson, et al. Development and Validation of a Scalable Next-Generation Sequencing System for Assessing Relevant Somatic Variants in Solid Tumors. Neoplasia. 2015 Apr; 17(4): 385-399.</p>
        
    <h4 style="background-color:#0059b3;color:White">DISCLAIMERS</h4>
        <p> 1. Next generation sequencing can identify somatic mutations, copy number aberrations and fusions present in the tumor specimen. However a negative test does not exclude the presence of mutated cells. Results of this test must always be interpreted in the context of clinical, morphologic and immunophenotypic data</p>
        <p> 2. Diagnostic sensitivity:</p>
        <p>&nbsp;&nbsp;&nbsp; a.This assay is designed to detect single nucleotide variants (SNV) only within defined regions.</p>
        <p>&nbsp;&nbsp;&nbsp; b.The assay detects amplifications (CNV) and fusions. Large deletions or large insertions (> 15 bp) may not be detected. </p>
        <p> 3.Analytical sensitivity:</p>
        <p>&nbsp;&nbsp;&nbsp; a.This assay may not detect certain SNV mutations if the proportion of tumor cells in the sample studied is less than 20%. The lower limit of detection (variant frequency) for SNVs, and indels is 0.05 and 0.07, respectively.</p>
        <p>&nbsp;&nbsp;&nbsp; b. The assay may not detect the CNV if the tumor content is less than 50%. Only high level amplifications (CN gains ≥ 5) are reported. </p>
        <p>&nbsp;&nbsp;&nbsp; c. Fusions may not be detected if the proportion of tumor cells in the sample studied is less than 20%. The lower limit of detection for fusions is between 10-2 and 10-3 based on our validation. Sub clonal amplifications present only in a subset of tumor cells may not be detected by this method.</p>
        <p> 4. Variants may not be detected in regions with coverage less than 500X and are therefore interpreted as indeterminate. A full list of those regions is available upon request.</p>
        <p> 5. Results of this test are not sufficient for diagnosis and must always be interpreted in the context of clinical, morphologic and immunophenotypic data.  </p>
        <p> 6. This test is not designed for detection of germline mutations. Genetic counselling and testing is necessary for analysis of germline mutations that may be associated with cancer</p>
        <p> 7. This test is not designed for detection of copy number losses in tested oncogenes.</p>
        <p> 8. This test was developed and its performance characteristics determined by the Molecular Pathology Laboratory of New York University. It has not been cleared or approved by the U.S. Food and Drug Administration. The FDA has determined that such clearance or approval is not necessary.</p>
        <br/>
    </div>
        
    <button onclick="Export2Doc('source-html');">Export to word doc</button><br/>
        
    <form method="POST" action="/index/{{runid}}"> 
        {% csrf_token %}
        <input type="submit" value="Home"/>
    </form>
              
    </body>
</html>
        
<script type="text/javascript">
    function Export2Doc(element, filename=''){
        var html = "<html xmlns:o='urn:schemas-microsoft-com:office:office' xmlns:w='urn:schemas-microsoft-com:office:word' xmlns='http://www.w3.org/TR/REC-html40'><head><meta charset='utf-8'></head><body>";
        var footer = "</body></html>";
        var html = html+document.getElementById(element).innerHTML+footer;
        
        //link url
        var url = 'data:application/vnd.ms-word;charset=utf-8,' + encodeURIComponent(html);
        //filename
        var filename_tumor = document.getElementById("tumorsample").innerHTML;
        var result = filename_tumor.split(": ")[1];
        var filename = result+'.doc';
        
        // Creates the  download link element dynamically
        var downloadLink = document.createElement("a");
        
        document.body.appendChild(downloadLink);
        
        //Link to the file
        downloadLink.href = url;
          
        //Setting up file name
        downloadLink.download = filename;
          
        //triggering the function
        downloadLink.click();
        //Remove the a tag after donwload starts.
        document.body.removeChild(downloadLink);
            
    }
</script>